File: src/mod11_gringotts_concurrency/lesson14_completable_future/java/Main.java
package mod11_gringotts_concurrency.lesson14_completable_future.java;

import java.util.concurrent.*;

//ForkJoinPool fork() join()
//RecursiveTask RecursiveAction

public class Main {

    public static final int CNT_TRANSACTIONS = 200;

    public static void main(String[] args) {

        BankAccount ollivander = new BankAccount("Ollivander's account", 15_000);
        BankAccount hagrid = new BankAccount("Hadrid's account", 500_000);

        //Future
//        ExecutorService executor = Executors.newFixedThreadPool(3);
//        Future<?> future1 = executor.submit(new Thread1(hagrid, ollivander));
//        Future<?> future2 = executor.submit(new RunnableForThread2(hagrid, ollivander));
//
//        try {
//            future1.get();
//            future2.get();
//        } catch (InterruptedException|ExecutionException e) {
//            throw new RuntimeException(e);
//        }
//
//        Future<String> futureTotal = executor.submit(() -> "Balances after transfers:" + hagrid.getAccountName() + ":" + hagrid.getBalance() + ";" +
//                ollivander.getAccountName() + ":" + ollivander.getBalance());
//
//        executor.shutdown();
//        try {
//            executor.awaitTermination(60, TimeUnit.SECONDS);
//            System.out.println(futureTotal.get());
//        } catch (InterruptedException | ExecutionException e) {
//            throw new RuntimeException(e);
//        }
        
        //CompletableFuture
        ExecutorService executor = Executors.newFixedThreadPool(3);

        CompletableFuture<Void> future1 = CompletableFuture.runAsync(new Thread1(hagrid, ollivander), executor);
        CompletableFuture<Void> future2 = CompletableFuture.runAsync(new RunnableForThread2(hagrid, ollivander), executor);

        CompletableFuture<Void> allOf = CompletableFuture.allOf(future1, future2);

        CompletableFuture<String> futureTotal = allOf.thenApplyAsync(
                (v) -> "Balances after transfers:" + hagrid.getAccountName() + ":" + hagrid.getBalance() + ";" +
                        ollivander.getAccountName() + ":" + ollivander.getBalance());

        futureTotal.thenAccept(System.out::println).join();
        executor.shutdown();

    }
}

File: src/mod11_gringotts_concurrency/lesson14_completable_future/kotlin/Main.kt
package mod11_gringotts_concurrency.lesson15_virtual_threads.kotlin

import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.Future

const val CNT_TRANSFERS = 200

fun main() {


    val hagridAccount = BankAccount("Hadrid's account", 500_000)
    val ollivanderAccount = BankAccount("Ollivander's account", 15_000)

    //Future
//    val executor = Executors.newFixedThreadPool(2)
//    val future1= executor.submit {
//        repeat(100) {
//            hagridAccount.transferFunds(ollivanderAccount, 10)
////            Transfer(hagridAccount,ollivanderAccount,10).execute()
//        }
//    }
//    val future2= executor.submit {
//        repeat(100) {
//            Transfer(hagridAccount,ollivanderAccount,10).execute()
//        }
//    }
//    future1.get()
//    future2.get()
//
//    executor.shutdown()
//
//    println(
//        "Balances after transfers:" + hagridAccount.accountName + ":" + hagridAccount.balance + ";" +
//                ollivanderAccount.accountName + ":" + ollivanderAccount.balance
//    )

    //CompletableFuture
    val executor = Executors.newFixedThreadPool(2)
    val future1= CompletableFuture.runAsync {
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
//            Transfer(hagridAccount,ollivanderAccount,10).execute()
        }
    }
    val future2= CompletableFuture.runAsync {
        repeat(100) {
            Transfer(hagridAccount,ollivanderAccount,10).execute()
        }
    }

    val allOf=CompletableFuture.allOf(future1,future2)

    val totalFuture=allOf.thenRunAsync {
        println(
            "Balances after transfers:" + hagridAccount.accountName + ":" + hagridAccount.balance + ";" +
                    ollivanderAccount.accountName + ":" + ollivanderAccount.balance
        )
    }.join()

    executor.shutdown()



}